File: interviews/leetcode/arrays-and-strings/1071. Greatest Common Divisor of Strings/euclidean_trick.py
import math
class Solution:
    # The trick is to first find the GCD of the two strings and
    # only use that length
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        # compute lengths
        l1 = len(str1)
        l2 = len(str2)

        # find the gcd of the lengths
        #gcd = math.gcd(l1, l2)
        gcd = self.gcd(l1, l2)

        # verify that both strings are repeated patterns
        pattern = str1[:gcd]
        pat2 = str2[:gcd]

        # both patterns must be same
        if pat2!=pattern:
            return ""

        # verify the strings only have one pattern
        for i in range(l1//gcd):
            if pattern!=str1[i*gcd:(i+1)*gcd]:
                return ""

        for i in range(l2//gcd):
            if pattern!=str2[i*gcd:(i+1)*gcd]:
                return ""

        return pattern

    def gcd(self, a: int, b:int):
        while b != 0:
            a, b = b, a%b
        return a

    def gcd_recursive(self, a, b):
        if(b == 0):
            return a
        return self.gcd_recursive(b, a % b)

File: interviews/leetcode/leetcode-75/arrays-and-strings/1071. Greatest Common Divisor of Strings/euclidean_trick.go
package main

func gcdOfStrings(str1 string, str2 string) string {
	// compute length of both
	l1 := len(str1)
	l2 := len(str2)

	// find the gcd of both lengths
	patLen := gcd(l1, l2)

	// find the pattern of both strings
	// this is the length of the shared pattern
	pat := str1[:patLen]

	// compare patterns
	if pat != str2[:patLen] {
		return ""
	}

	// make sure str1 doesn't have any weird pattern
	for i := 0; i < len(str1)/patLen; i++ {
		if pat != str1[i*patLen:(i+1)*patLen] {
			return ""
		}
	}

	// make sure str1 doesn't have any weird pattern
	for i := 0; i < len(str2)/patLen; i++ {
		if pat != str2[i*patLen:(i+1)*patLen] {
			return ""
		}
	}

	// return successful pattern
	return pat
}

func gcd(a, b int) int {
	// base case
	if a == 0 || b == 0 {
		return 0
	}
	// generic case
	for b != 0 {
		a, b = b, a%b
	}
	return a
}
